src/domain/swaps/contracts/decoders/gp-v2-decoder.helper.ts

```diff
@@ -21,16 +654,115 @@ export class GPv2Decoder extends AbiDecoder<typeof abi> {
    * @param data - the transaction data for the setPreSignature call
    * @returns {`0x${string}`} the order UID or null if the data does not represent a setPreSignature transaction
    */
-  getOrderUidFromSetPreSignature(data: `0x${string}`): `0x${string}` | null {
+  public getOrderUidFromSetPreSignature(
+    data: `0x${string}`,
+  ): `0x${string}` | null {
+    if (!this.helpers.isSetPreSignature(data)) {
+      return null;
+    }
+
     try {
-      if (!this.helpers.isSetPreSignature(data)) return null;
-      const { args } = this.decodeFunctionData({ data });
-      return args[0];
+      const decoded = this.decodeFunctionData({ data });
+
+      if (decoded.functionName !== 'setPreSignature') {
+        throw new Error('Data is not of setPreSignature');
+      }
+
+      return decoded.args[0];
     } catch (e) {
-      this.loggingService.debug(e);
       return null;
     }
   }
+
+  /**
+   * Decodes an order from a settlement trade.
+   *
+   * @param trade The trade to decode into an order.
+   * @param tokens The list of token addresses as they appear in the settlement.
+   * @returns The decoded {@link GPv2OrderParameters} or null if the trade is invalid.
+   */
+  public decodeOrderFromSettlement(
+    data: `0x${string}`,
+  ): GPv2OrderParameters | null {
+    const decoded = this.decodeSettle(data);
+
+    if (!decoded) {
+      return null;
+    }
+
+    const [tokens, , [trade]] = decoded;
+
+    const sellTokenIndex = Number(trade.sellTokenIndex);
+    const buyTokenIndex = Number(trade.buyTokenIndex);
+
+    if (Math.max(sellTokenIndex, buyTokenIndex) >= tokens.length) {
+      throw new Error('Invalid trade');
+    }
+
+    return {
+      sellToken: tokens[sellTokenIndex],
+      buyToken: tokens[buyTokenIndex],
+      receiver: trade.receiver,
+      sellAmount: trade.sellAmount,
+      buyAmount: trade.buyAmount,
+      validTo: trade.validTo,
+      appData: trade.appData,
+      feeAmount: trade.feeAmount,
+      kind: this.decodeFlag('kind', trade.flags),
+      partiallyFillable: this.decodeFlag('partiallyFillable', trade.flags),
+      sellTokenBalance: this.decodeFlag('sellTokenBalance', trade.flags),
+      buyTokenBalance: this.decodeFlag('buyTokenBalance', trade.flags),
+    };
+  }
+
+  // Use inferred return type
+  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+  private decodeSettle(data: `0x${string}`) {
+    if (!this.helpers.isSettle(data)) {
+      return null;
+    }
+
+    try {
+      const decoded = this.decodeFunctionData({ data });
+
+      if (decoded.functionName !== 'settle') {
+        throw new Error('Data is not of settle');
+      }
+
+      return decoded.args;
+    } catch (e) {
+      return null;
+    }
+  }
+
+  private decodeFlag<K extends keyof typeof GPv2Decoder.FlagMasks>(
```

Comment:
Could we include some doc on this function? 🙏 It's not super trivial

Reply:
Added in a3e9d98e906fac4a548304dbcba9544e68264764. Please lmk if it is sufficient.